### PR TITLE
Turned Views, Layouts, and Loops into proper dataclasses

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137861
* #137410
* #138788
* #138891
* #138787
* #138881
* #138794
* #138733

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by Skylion007** [comment]
FYI: In case this bites you: https://github.com/python/cpython/issues/109409

**Comment by alexdremov** [comment]
https://github.com/pytorch/pytorch/blob/03de794c34538f9cd4587ea4097486ef14e10840/torch/_inductor/ir.py#L5203

also should have kwargs for appropriate dataclass init

**Comment by alexdremov** [comment]
Hey, @Chillee! I'm from the issue #136196

I'm catching [this runtime error](https://github.com/pytorch/pytorch/blob/03de794c34538f9cd4587ea4097486ef14e10840/torch/_higher_order_ops/flex_attention.py#L90) that should be relevant to your changes. 

After debug prints, I see that the buffers in the assertion have the type of `SymInt`

**Comment by Chillee** [comment]
@alexdremov ah I don't think that's related, i think it's another limitation related to symbolic shapes today.

**Comment by alexdremov** [comment]
> @alexdremov ah I don't think that's related, i think it's another limitation related to symbolic shapes today.

Are there quick fixes? 

If no, the initial issue should be left open as combination of dynamic shapes + flex attention is not working as of this PR.

**Comment by Chillee** [comment]
@alexdremov The example in that particular issue works, right? 

The error you're running into is most likely because you're doing something like
```
Q_LEN = q.shape[0]
def score_mod(score, b, h, q, kv):
    return score + (Q_LEN - kv)
```
This can be worked around by converting `Q_LEN` into a tensor. 

There's some discussion here https://github.com/pytorch/pytorch/issues/137779

**Comment by alexdremov** [comment]
> This can be worked around by converting `Q_LEN` into a tensor.

I see, I'll try doing this thing.

My example is nothing fancy. It's just sequence length padding + chunking

```python
# pseudocode-ish

lens = torch.ones((B, ), device='cuda').long() # sequence lengths

def generator(b, h, q_idx, kv_idx):
    chunking = f(q_idx, kv_idx)  # only simple torch operations inside f
    padding = (q_idx < lens[b]) & (kv_idx < lens[b])
    return chunking & padding

# for self attention
block_mask = create_block_mask(generator, B=B, H=None, Q_LEN= Q_LEN, KV_LEN= KV_LEN)
```
